_overviews/scala3-book/domain-modeling-tools.md

```diff
@@ -0,0 +1,575 @@
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+To use them in other code, first import them, and then use them:
+
+```scala
+import CrustSize._
+val currentCrustSize = Small
+```
+
+They can be used anywhere classes can be used, including if/then expressions, match expressions, and more:
+
+```scala
+// if/then
+if (currentCrustSize == Large)
+  println("You get a prize!")
+
+// match
+currentCrustSize match
+  case Small => println("small")
+  case Medium => println("medium")
+  case Large => println("large")
+```
+
+In the modeling sections that follow, you’ll also see that they can be used as constructor parameters, and to model ADTs and GADTs.
+
+### More enum features
+
+Enumerations have more features. You have the basic enum just shown:
+
+```scala
+enum Color:
+  case Red, Green, Blue
+```
+
+They can also be parameterized:
+
+```scala
+enum Color(val rgb: Int):
+  case Red   extends Color(0xFF0000)
+  case Green extends Color(0x00FF00)
+  case Blue  extends Color(0x0000FF)
+```
+
+And they can also have user-defined members:
+
+```scala
+enum Planet(mass: Double, radius: Double):
+  private final val G = 6.67300E-11
+  def surfaceGravity = G * mass / (radius * radius)
+  def surfaceWeight(otherMass: Double) =  otherMass * surfaceGravity
+
+  case Mercury extends Planet(3.303e+23, 2.4397e6)
+  case Venus   extends Planet(4.869e+24, 6.0518e6)
+  case Earth   extends Planet(5.976e+24, 6.37814e6)
+  // more planets here ...
+```
+
+### Compatibility with Java Enums
+
+If you want to use Scala-defined enums as Java enums, you can do so by extending the class *java.lang.Enum*, which is imported by default, as follows:
+
+```scala
+enum Color extends Enum[Color] { case Red, Green, Blue }
+```
+
+The type parameter comes from the Java `enum` definition, and should be the same as the type of the enum. There’s no need to provide constructor arguments (as defined in the Java API docs) to *java.lang.Enum* when extending it — the compiler generates them automatically.
+
+After defining `Color` like that, you can use it like you would a Java enum:
+
+````
+scala> Color.Red.compareTo(Color.Green)
+val res0: Int = -1
+````
+
+
+
+## Case classes
+
+A `case class` has all of the functionality of a `class`, and more. When the compiler sees the `case` keyword in front of a `class`, it generates code for you, with the following benefits:
+
+{% comment %}
+TODO: What to say about `apply` methods in the following bullet points? They’re less significant in Scala 3.
+{% endcomment %}
+
+* Case class constructor parameters are public `val` fields by default, so accessor methods are generated for each parameter
+* An `apply` method is created in the companion object of the class, so you don’t need to use the `new` keyword to create a new instance of the class
+* An `unapply` method is generated, which lets you use case classes in more ways in `match` expressions
```

Comment:
IIRC the requirements for pattern matching changed in scala 3 as well

Reply:
Yes, there is this document describing "Optionless" pattern matching: http://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html

Reply:
In trying to resolve this, is there something specific we should add here? Or should we create another case class discussion elsewhere?

Reply:
I think we could discuss it here, but I would defer this to a revisiin, after we release the docs